### PR TITLE
Treat warnings as errors

### LINKED ISSUE #374: treat warnings as errors
https://github.com/ry/deno/blob/ed9acaec7d438515271c1f2664780c0612f339da/.gn#L28
The task is to delete this line and deal with all the warnings that cause the build to fail. [There are a handful.](https://travis-ci.com/ry/deno/builds/79153409) Bonus points if you make rustc warnings also obey the `treat_warnings_as_errors` gn arg.

### PR DESCRIPTION
**Pull request by kt3k**
This resolves #374.

This PR depends on #371. I'm going to rebase it if #371 gets merged.

### CODE REVIEW

File: build_extra/flatbuffers/BUILD.gn [hunk 1]
```diff
@@ -13,6 +13,7 @@ config("flatbuffers_config") {
     cflags = [
       "-Wno-exit-time-destructors",
       "-Wno-header-hygiene",
+      "-Wno-return-type", # TODO: rust branch of flatbuffers has this warning. this should be removed when the branch fixed.
```
Comment (ry) [inline build_extra/flatbuffers/BUILD.gn]: Wrap at 80 characters. `./tools/format.py` should take care of this.

File: build_extra/rust/rust.gni [hunk 2]
```diff
@@ -40,6 +40,7 @@ template("run_rustc") {
 
     args = [
       rebase_path(source_root, root_build_dir),
+      "-Dwarnings",
```
Comment (ry) [inline build_extra/rust/rust.gni]: Nice. Can you add this conditionally based on treat_warnings_as_errors ?

File: build_extra/rust/BUILD.gn [hunk 3]
```diff
@@ -32,6 +32,11 @@ rust_component("url") {
 
 rust_component("percent_encoding") {
   source_root = "$crates/url/percent_encoding/lib.rs"
+  args = [
+    # suppress some warnings
+    "-Aunused-imports",
+    "-Adeprecated",
+  ]
```
Comment (ry) [inline build_extra/rust/BUILD.gn]: Why does this module have these warnings? Do the authors not care (seems unlikely) or are we compiling it with the wrong cfg? Can you investigate further? If it's their issue, please open an issue with them and link to the issue here.

File: build_extra/rust/rust.gni [hunk 4]
```diff
@@ -102,6 +103,10 @@ template("run_rustc") {
       }
     }
 
+    if (defined(invoker.args)) {
+      args += invoker.args
+    }
```
Comment (ry) [inline build_extra/rust/rust.gni]: I'd prefer not to add this... but it's fine if you can't fix the warnings in `percent_encoding`

**Comment by ry** [comment]
Great! Thank you!

File: build_extra/rust/BUILD.gn (hunk 3 above)
Comment (kt3k) [reply build_extra/rust/BUILD.gn]: They seem caused by the line `use std::ascii::AsciiExt;`. The module seems deprecated at recent version of rust. Their build of master branch also shows [the same warnings](https://travis-ci.org/servo/rust-url/jobs/400867523). So I think it's not a problem of our configuration.

I think this should be fixed in their repository. I filed an issue there. https://github.com/servo/rust-url/issues/455

File: .gn [hunk 5]
```diff
@@ -25,7 +25,7 @@ default_args = {
 
   is_component_build = false
   symbol_level = 1
-  treat_warnings_as_errors = false
+  treat_warnings_as_errors = true
```
Comment (ry) [inline .gn]: Remove this line?

File: build_extra/rust/BUILD.gn [hunk 6]
```diff
@@ -34,6 +34,12 @@ rust_component("url") {
 
 rust_component("percent_encoding") {
   source_root = "$crates/registry/src/github.com-1ecc6299db9ec823/percent-encoding-1.0.1/lib.rs"
+  args = [
+    # TODO: Suppress some warnings at this moment
+    # This should be removed when it's fixed in servo/rust-url repository
+    "-Aunused-imports",
+    "-Adeprecated",
```
Comment (ry) [inline build_extra/rust/BUILD.gn]: Link to the issue you created here

**Comment by kt3k** [comment]
I reverted the following part because it doesn't seem working.
```gn
    if (defined(treat_warnings_as_errors) && treat_warnings_as_errors) {
      args += [ "-Dwarnings" ]
    }
```
 (`defined(treat_warnings_as_errors)` seems evaluated always `false`). I couldn't find a way to use that build arg in this scope. I left a `TODO` comment instead.

**Comment by ry** [comment]
LGTM

@piscisaureus please check this works for you.

**Comment by piscisaureus** [comment]
It doesn't work for me. On linux I get a lot of errors like this:

```
clang++: error: argument unused during compilation: '-isystem ../../buildtools/third_party/libc++abi/trunk/include' [-Werror,-Wunused-command-line-argument]
```

On windows it breaks here:
```
C:/programming/depot_tools/win_tools-2_7_6_bin/python/bin/python.exe ../../tools/run_rustc.py ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/libc-0.2.42/src/lib.rs --crate-name=libc --crate-type=rlib -Dwarnings --emit=link=obj/build_extra/rust/liblibc.rlib --emit=dep-info=obj/build_extra/rust/libc.d --depfile=obj/build_extra/rust/libc.d --output_file=obj/build_extra/rust/liblibc.rlib -g --cfg "feature=\"use_std\""
error: unused macro definition
  --> ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/libc-0.2.42/src\macros.rs:51:1
   |
51 | / macro_rules! f {
52 | |     ($(pub fn $i:ident($($arg:ident: $argty:ty),*) -> $ret:ty {
53 | |         $($body:stmt);*
54 | |     })*) => ($(
...  |
66 | |     )*)
67 | | }
   | |_^
   |
   = note: `-D unused-macros` implied by `-D warnings`
```

**Comment by kt3k** [comment]
@piscisaureus 
Thanks for checking! I'll investigate them.

**Comment by piscisaureus** [comment]
@kt3k I have to warn you, I do get a lot more warning, especially on linux. 
However any warning you can fix is great!

```
piscisaureus@pumpkin:~/d/deno$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.3 LTS
Release:        16.04
Codename:       xenial
```

**Comment by ry** [comment]
@kt3k if it gets difficult to keep up with rebases we can land a partial fix now with these changes but without turning on `treat_warnings_as_errors`.

**Comment by kt3k** [comment]
The build args in `.travis.yml` (`use_custom_libcxx=false use_sysroot=false`) seem suppressing a lot of warnings (such as `unused-command-line-argument`) on ubuntu, but I'm still seeing many...

I'm going to try some more variations for a while.

**Comment by kt3k** [comment]
@ry @piscisaureus 
I reverted the arg `treat_warnings_as_errors` for now. I'd like to merge at this point.

When I set `use_allocator="none" and is_clang=false` on ubuntu (16.04) as well as the above args, then it reduces the warnings further but it causes different warnings in deno's source code, such as `unused-variable` and `unused-result`. If I try to suppress those, maybe I need to write g++ specific pragmas, but I'm not sure that's the right direction to go.

**Pull request merged**